src/json.hpp

```diff
@@ -8222,33 +8224,43 @@ class basic_json
 
     @since version 1.0.0
     */
-    class const_iterator : public std::iterator<std::random_access_iterator_tag, const basic_json>
+    template <typename U>
+    class iter_impl : public std::iterator<std::random_access_iterator_tag, U>
     {
         /// allow basic_json to access private members
         friend class basic_json;
 
+        // make sure U is basic_json or const basic_json
+        static_assert(std::is_same<U, basic_json>::value
+                      or std::is_same<U, const basic_json>::value,
+                      "iter_impl only accepts (const) basic_json");
+
```

Comment:
As the iterators are only defined inside the `basic_json` class, I think it's OK to skip the error message and move it to the specification. The code could maybe even simplified by using `std::is_same<U, std::remove_cv<basic_json>>`.